### PR TITLE
SAS token API does not work with storage containers created in Landing Zone storage accounts

### PR DESCRIPTION
**Pull request by sergiygetlin**
https://broadworkbench.atlassian.net/browse/TOAZ-264
-Fix issue: handle situation when storage container belongs to shared storage account from landing zone
-Refactoring

### CODE REVIEW

File: service/src/main/java/bio/terra/workspace/app/controller/ControlledAzureResourceApiController.java [hunk 1]
```diff
@@ -73,7 +74,7 @@
 @Controller
 public class ControlledAzureResourceApiController extends ControlledResourceControllerBase
     implements ControlledAzureResourceApi {
-  private final Logger logger = LoggerFactory.getLogger(ControlledGcpResourceApiController.class);
```
Comment (zykonda) [inline service/src/main/java/bio/terra/workspace/app/controller/ControlledAzureResourceApiController.java:76]: This is good catch :)

File: service/src/test/java/bio/terra/workspace/service/resource/controlled/cloud/azure/CreateAndDeleteAzureControlledResourceFlightTest.java [hunk 2]
```diff
@@ -497,6 +489,180 @@ public void createAzureStorageContainerFlightFailedBecauseLandingZoneDoesntExist
...
+    TimeUnit.MINUTES.sleep(1);
+
+    // Submit a storage container creation flight and then verify the resource exists in the
+    // workspace.
+    final UUID containerResourceId = UUID.randomUUID();
+    final String storageContainerName = ControlledResourceFixtures.uniqueBucketName();
+    ControlledAzureStorageContainerResource containerResource =
+        ControlledAzureStorageContainerResource.builder()
+            .common(
+                ControlledResourceFields.builder()
```
Comment (zykonda) [inline service/src/test/java/bio/terra/workspace/service/resource/controlled/cloud/azure/CreateAndDeleteAzureControlledResourceFlightTest.java:623]: nit: building ControlledResourceFields object is repeated several times, may be taken to a private method or test fixture method.

**Review by zykonda — APPROVED** [review]
Design changes make sense, aligned with Azure resource and landing zone concepts, code is cleaner now.
Looks good to me.

File: service/src/main/java/bio/terra/workspace/service/resource/controlled/cloud/azure/AzureStorageAccessService.java [hunk 3]
```diff
@@ -247,4 +263,59 @@ public BlobContainerClient buildBlobContainerClient(
         .containerName(containerResource.getStorageContainerName())
         .buildClient();
   }
+
+  private StorageData getStorageAccountData(
```
Comment (cahrens) [inline service/src/main/java/bio/terra/workspace/service/resource/controlled/cloud/azure/AzureStorageAccessService.java:267]: There is similar code in https://github.com/DataBiosphere/terra-workspace-manager/blob/688c3f42db57fcb1206f2a3c3a45ff6d1d12f85e/service/src/main/java/bio/terra/workspace/service/resource/controlled/cloud/azure/storageContainer/VerifyAzureStorageContainerCanBeCreatedStep.java (see `validateStorageAccountExists`). Can we have a shared method that gets the parent storage account of a container?

File: service/src/test/java/bio/terra/workspace/service/resource/controlled/cloud/azure/AzureStorageAccessServiceUnitTest.java [hunk 4]
```diff
@@ -422,19 +424,128 @@ void createAzureStorageContainerSasToken_blobName() throws InterruptedException
...
-            null,
-            "foo/the/bar.baz",
-            null);
+            new SasTokenOptions(null, startTime, expiryTime, "foo/the/bar.baz", null));
 
     assertValidToken(result.sasToken(), BlobContainerSasPermission.parse("racwdl"), true);
   }
+
+  @Test
+  public void createAzureStorageContainerSasToken_basedOnLzSharedStorageAccount_Success()
```
Comment (aherbst-broad) [inline service/src/test/java/bio/terra/workspace/service/resource/controlled/cloud/azure/AzureStorageAccessServiceUnitTest.java:440]: Is this named correctly? This test appears to be asserting that an IP address is added to the sas.

File: service/src/test/java/bio/terra/workspace/service/resource/controlled/cloud/azure/AzureStorageAccessServiceUnitTest.java [hunk 5]
```diff
@@ -422,19 +424,128 @@ void createAzureStorageContainerSasToken_blobName() throws InterruptedException
...
+  public void createAzureStorageContainerSasToken_LzSharedStorageAccountNotFound_Failure()
+      throws InterruptedException {
+    var storageContainerResource =
+        buildStorageContainerResource(
+            PrivateResourceState.NOT_APPLICABLE,
+            AccessScopeType.ACCESS_SCOPE_SHARED,
+            ManagedByType.MANAGED_BY_USER,
+            null);
+
+    var ipRange = "168.1.5.60-168.1.5.70";
```
Comment (aherbst-broad) [inline service/src/test/java/bio/terra/workspace/service/resource/controlled/cloud/azure/AzureStorageAccessServiceUnitTest.java:484]: Why add the ip range in this test?

File: service/src/test/java/bio/terra/workspace/service/resource/controlled/cloud/azure/CreateAndDeleteAzureControlledResourceFlightTest.java (hunk 2 above)
Comment (aherbst-broad) [reply service/src/test/java/bio/terra/workspace/service/resource/controlled/cloud/azure/CreateAndDeleteAzureControlledResourceFlightTest.java:623]: [ControlledResourceFixtures](https://github.com/DataBiosphere/terra-workspace-manager/blob/main/service/src/test/java/bio/terra/workspace/common/fixtures/ControlledResourceFixtures.java) might a good home for such code.

**Review by aherbst-broad — APPROVED** [review]
Thanks for doing this, had a few comments.

File: service/src/test/java/bio/terra/workspace/service/resource/controlled/cloud/azure/AzureStorageAccessServiceUnitTest.java (hunk 5 above)
Comment (sergiygetlin) [reply service/src/test/java/bio/terra/workspace/service/resource/controlled/cloud/azure/AzureStorageAccessServiceUnitTest.java:484]: It is just an option for SAS token. It is not specific to this test. We can remove it or use another options.

File: service/src/test/java/bio/terra/workspace/service/resource/controlled/cloud/azure/AzureStorageAccessServiceUnitTest.java (hunk 4 above)
Comment (sergiygetlin) [reply service/src/test/java/bio/terra/workspace/service/resource/controlled/cloud/azure/AzureStorageAccessServiceUnitTest.java:440]: Here we are testing token generation for storage container related to shared storage account. I used one of the parameters for token generation just for test purpose. So, to check that we still have correct token. It is possible to skip  all those parameters as well.

File: service/src/test/java/bio/terra/workspace/service/resource/controlled/cloud/azure/AzureStorageAccessServiceUnitTest.java (hunk 5 above)
Comment (aherbst-broad) [reply service/src/test/java/bio/terra/workspace/service/resource/controlled/cloud/azure/AzureStorageAccessServiceUnitTest.java:484]: Got it; yeah I'd suggest removing it.

File: service/src/main/java/bio/terra/workspace/service/resource/controlled/cloud/azure/AzureStorageAccessService.java (hunk 3 above)
Comment (sergiygetlin) [reply service/src/main/java/bio/terra/workspace/service/resource/controlled/cloud/azure/AzureStorageAccessService.java:267]: Agree. Code is similar in some way. It's probably good place for improvement. I think that this improvement might change the scope of current ticket. I created separate Jira ticket for this - TOAZ-278.

**Pull request merged**